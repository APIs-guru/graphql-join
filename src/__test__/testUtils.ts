declare var expect;

import {
  Source,
  DocumentNode,
  GraphQLError,
  GraphQLSchema,
  GraphQLFieldResolver,
  GraphQLResolveInfo,
  ExecutionArgs,
  ExecutionResult,

  graphql,
  parse,
  validate,
  execute,
  print,
  buildSchema,
  printSchema,
  formatError,
  specifiedRules,
} from 'graphql';
import * as _ from 'lodash';

import { GraphQLJoinSchema, ProxyContext } from '../index';
import { stubSchema } from '../utils';

expect.addSnapshotSerializer({
  print(val) {
    return typeof val === 'string' ?
      val.trim() :
      JSON.stringify(val, null, 2);
  },
  test(val) {
    return true;
  },
});

function fakeFieldResolver(schemaName: string): GraphQLFieldResolver<any,any> {
  return (
    _1, args: object, _3,
    {fieldName, parentType}: GraphQLResolveInfo
  ) => {
    let result = `${schemaName}::${parentType.name}::${fieldName}`;
    if (Object.keys(args).length !== 0) {
      result += `;args=${JSON.stringify(args)}`
    }
    return result;
  }
}

function makeProxy(schemaName: string, schema: GraphQLSchema) {
  return (queryAST: DocumentNode, variableValues?: object) => {
    const query = print(queryAST);
    expect(query).toMatchSnapshot();
    // FIXME: Should use executeQuery but blocked by
    // https://github.com/facebook/jest/issues/3917
    return graphql({
      schema,
      source: new Source(query, 'Send to ' + schemaName),
      variableValues,
      fieldResolver: fakeFieldResolver(schemaName),
    });
  };
}

type TestSchema = string | { idl: string, prefix?: string };
type TestSchemasMap = { [name: string]: TestSchema };
export function testJoin(testSchemas: TestSchemasMap, joinIDL: string) {
  const remoteSchemas = _.mapValues(testSchemas, (schemaSource, name) => {
    if (typeof schemaSource === 'string') {
      schemaSource = { idl: schemaSource };
    }
    const { prefix, idl } = schemaSource;
    const schema = buildSchema(new Source(idl, name));
    stubSchema(schema);
    return { schema , prefix};
  });

  const joinSchema = new GraphQLJoinSchema(
    new Source(joinIDL, 'Join SDL'),
    remoteSchemas
  );
  const schema = joinSchema.schema;

  expect(schema).toBeInstanceOf(GraphQLSchema);
  expect(printSchema(schema)).toMatchSnapshot();
  return async (
    query: string | { query: string, variableValues: object },
    results?: { [schemaName: string]: ExecutionResult }
  ) => {
    const proxyFns = _.mapValues(remoteSchemas, ({schema}, name) => {
      if (results && results[name]) {
        return (queryAST) => {
          const query = print(queryAST);
          expect(query).toMatchSnapshot();
          return Promise.resolve(results[name])
        };
      }
      return makeProxy(name, schema)
    });

    const queryObj = typeof query === 'string' ?
      { query: query, variableValues: {} } : query;
    const result = await executeQuery({
      schema,
      document: parse(new Source(queryObj.query, 'ClientQuery')),
      variableValues: queryObj.variableValues,
      contextValue: new ProxyContext(joinSchema, proxyFns),
    });

    expect([
      query,
      resultToJSON(result),
    ]).toMatchSnapshot();
  };
}

function executeQuery(args: ExecutionArgs) {
  const validationErrors = validate(args.schema, args.document, specifiedRules);
  expect(validationErrors.join('\n')).toBe('');
  return execute(args);
}

function resultToJSON(result) {
  return {
    ...result,
    errors: (result.errors || []).map(formatError),
  };
}

