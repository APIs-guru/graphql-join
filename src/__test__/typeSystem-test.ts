declare var describe, test;

import { testJoin } from './testUtils';

describe('type system', () => {

  test('arrays in response type', async () => {
    const execute = testJoin({
      test: `
        type Query { foo: [[Bar]] }
        type Bar { bar: String }
      `,
    }, `
      type Query {
        foo: [[String]] @resolveWith(query: "foo")
      }

      query foo @send(to: "test") {
        foo {
          bar
        }
      }
    `);

    await execute('{ foo }', { test: {
      data: {
        foo: [
          [{bar: 'a'}, {bar: 'b'}],
          [{bar: 'c'}]
        ],
      }
    }});
  });

  test('arrays of objects with arrays in response type', async () => {
    const execute = testJoin({
      test: `
        type Query { foo: [Bar] }
        type Bar { bar: [Baz] }
        type Baz { baz: String }
      `,
    }, `
      type Query {
        foo: [[String]] @resolveWith(query: "foo")
      }

      query foo @send(to: "test") {
        foo {
          bar {
            baz
          }
        }
      }
    `);

    await execute('{ foo }', { test: {
      data: {
        foo: [
          {
            bar: [{baz: 'a'}, {baz: 'b'}],
          },
          {
            bar: [{baz: 'c'}]
          },
        ],
      }
    }});
  });
});
