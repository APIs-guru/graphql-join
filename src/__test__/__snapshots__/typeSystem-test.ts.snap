// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`type system arrays in response type 1`] = `
type Query {
  foo: [[String]]
}
`;

exports[`type system arrays in response type 2`] = `
{
  foo {
    bar
  }
}
`;

exports[`type system arrays in response type 3`] = `
[
  "{ foo }",
  {
    "data": {
      "foo": [
        [
          "a",
          "b"
        ],
        [
          "c"
        ]
      ]
    },
    "errors": []
  }
]
`;

exports[`type system arrays of objects with arrays in response type 1`] = `
type Query {
  foo: [[String]]
}
`;

exports[`type system arrays of objects with arrays in response type 2`] = `
{
  foo {
    bar {
      baz
    }
  }
}
`;

exports[`type system arrays of objects with arrays in response type 3`] = `
[
  "{ foo }",
  {
    "data": {
      "foo": [
        [
          "a",
          "b"
        ],
        [
          "c"
        ]
      ]
    },
    "errors": []
  }
]
`;
